### PR TITLE
Extracting out unrelated logic from TestNG class.

### PR DESCRIPTION
**Pull request by krmahadevan**
* Moved some of the parsing logic to the Parser class.
* Moved some of the zip file parsing logic to its
own class.
* Added a test jar to support tests for Zip file
extraction logic.

Fixes # .

### Did you remember to?

- [ ] Add test case(s)
- [ ] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/JarFileUtils.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+            List<String> classes = Lists.newArrayList();
+            boolean foundTestngXml = testngXmlExistsInJar(jarFile, classes);
+            if (!foundTestngXml) {
+                Utils.log("TestNG", 1,
+                        "Couldn't find the " + xmlPathInJar + " in the jar file, running all the classes");
+                suites.add(XmlSuiteUtils.newXmlSuiteUsing(classes));
+            }
+        } catch (IOException ex) {
+            ex.printStackTrace();
```
Comment (juherr) [inline src/main/java/org/testng/JarFileUtils.java]: Maybe throw an TestNGException

File: src/main/java/org/testng/JarFileUtils.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
+                    for (XmlSuite suite : parsedSuites) {
+                        // If test names were specified, only run these test names
+                        XmlSuite suiteToAdd = suite;
+                        if (testNames != null) {
+                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
+                        }
+                        suites.add(suiteToAdd);
+                    }
+                    foundTestngXml = true;
```
Comment (juherr) [inline src/main/java/org/testng/JarFileUtils.java]: you can return directly

File: src/test/java/org/testng/JarFileUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import java.net.URLClassLoader;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class JarFileUtilsTest {
+    private static final File jar = new File("src/test/resources/testng-tests.jar");
```
Comment (juherr) [inline src/test/java/org/testng/JarFileUtilsTest.java]: Note for later: the jar should/could be generated by the test (for example with https://github.com/shrinkwrap/shrinkwrap)

**Review by juherr — APPROVED** [review]
Awesome change 👍

**Review by cbeust — COMMENTED** [review]
LGTM.

File: src/main/java/org/testng/JarFileUtils.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/JarFileUtils.java]: Fixed.

File: src/main/java/org/testng/JarFileUtils.java (hunk 2 above)
Comment (krmahadevan) [reply src/main/java/org/testng/JarFileUtils.java]: Fixed.

File: src/test/java/org/testng/JarFileUtilsTest.java (hunk 3 above)
Comment (krmahadevan) [reply src/test/java/org/testng/JarFileUtilsTest.java]: Added a TODO

**Comment by krmahadevan** [comment]
@cbeust - Thanks for approving the PR. Since I remember you had mentioned earlier, that I can go ahead and merge, if its been approved by you, I fixed the comments from @juherr and then merged it.

**Comment by cbeust** [comment]
Yup, go right ahead @krmahadevan. I noticed @juherr hadn't looked yet so I didn't want to merge it just yet but it looked good to me.

**Pull request merged**